src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -1937,53 +1823,45 @@ void PhaseIdealLoop::update_main_loop_assertion_predicates(Node* ctrl, CountedLo
                                                                      const Node_List& old_new) {
   const NodeInOriginalLoopBody node_in_original_loop_body(first_node_index_in_cloned_loop_body, old_new);
-  create_assertion_predicates_at_loop(peeled_loop_head, remaining_loop_head, node_in_original_loop_body);
+  create_assertion_predicates_at_loop(peeled_loop_head, remaining_loop_head, node_in_original_loop_body, false);
+}
+
+// Source Loop: Cloned   - pre_loop_head
+// Target Loop: Original - main_loop_head
+void PhaseIdealLoop::initialize_assertion_predicates_for_main_loop(CountedLoopNode* main_loop_head,
+                                                                   CountedLoopNode* pre_loop_head,
+                                                                   const uint first_node_index_in_cloned_loop_body,
+                                                                   const Node_List& old_new) {
+  const NodeInOriginalLoopBody node_in_original_loop_body(first_node_index_in_cloned_loop_body, old_new);
+  create_assertion_predicates_at_loop(pre_loop_head, main_loop_head, node_in_original_loop_body, true);
+}
+
+// Source Loop: Original - main_loop_head
+// Target Loop: Cloned   - post_loop_head
+void PhaseIdealLoop::initialize_assertion_predicates_for_post_loop(CountedLoopNode* main_loop_head,
+                                                                   CountedLoopNode* post_loop_head,
+                                                                   const uint first_node_index_in_cloned_loop_body) {
+  const NodeInClonedLoopBody node_in_cloned_loop_body(first_node_index_in_cloned_loop_body);
+  create_assertion_predicates_at_loop(main_loop_head, post_loop_head, node_in_cloned_loop_body, false);
 }
 
 void PhaseIdealLoop::create_assertion_predicates_at_loop(CountedLoopNode* source_loop_head,
-                                                         CountedLoopNode* target_loop_head,
-                                                         const NodeInLoopBody& _node_in_loop_body) {
+                                                          CountedLoopNode* target_loop_head,
+                                                          const NodeInLoopBody& _node_in_loop_body,
+                                                          const bool clone_template) {
   Node* init = target_loop_head->init_trip();
   Node* stride = target_loop_head->stride();
   LoopNode* target_outer_loop_head = target_loop_head->skip_strip_mined();
   Node* target_loop_entry = target_outer_loop_head->in(LoopNode::EntryControl);
   CreateAssertionPredicatesVisitor create_assertion_predicates_visitor(init, stride, target_loop_entry, this,
-                                                                       _node_in_loop_body);
+                                                                       _node_in_loop_body, clone_template);
```

Comment:
Not a very elegant solution to introduce `clone_template` but I'm planning to update this code again at a later point. This was the least invasive way to handle this.